src/cpufreq.c

```diff
@@ -39,23 +44,41 @@ struct feat_t {
     bool pcluster_only;
 };
 
+static uint32_t get_target_pstate(const struct cluster_t *cluster)
+{
+    u64 val = read64(cluster->base + CLUSTER_PSTATE);
+
+    if (chip_id == S5L8960X || chip_id == T7000 || chip_id == T7001)
+        return FIELD_GET(CLUSTER_PSTATE_DESIRED1_S5L8960X, val);
+    else
+        return FIELD_GET(CLUSTER_PSTATE_DESIRED1, val);
+}
+
 static int set_pstate(const struct cluster_t *cluster, uint32_t pstate)
 {
     u64 val = read64(cluster->base + CLUSTER_PSTATE);
 
-    if (FIELD_GET(CLUSTER_PSTATE_DESIRED1, val) != pstate) {
-        val &= ~CLUSTER_PSTATE_DESIRED1;
-        val |= CLUSTER_PSTATE_SET | FIELD_PREP(CLUSTER_PSTATE_DESIRED1, pstate);
-        if (chip_id == T8103 || chip_id <= T6002) {
+    if (get_target_pstate(cluster) != pstate) {
+        if (chip_id == S5L8960X || chip_id == T7000 || chip_id == T7001) {
```

Comment:
I'd just use a switch/case with all chip_id listed explicitly here even if that results in some code duplication. Even the chip_id <= T6002 in the original code is a bit weird but now that we have three(?) different variants this is getting confusing.